waspc/cli/src/Wasp/Cli/GithubRepo.hs

```diff
@@ -0,0 +1,129 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Wasp.Cli.GithubRepo where
+
+import qualified Codec.Archive.Tar as Tar
+import qualified Codec.Compression.GZip as GZip
+import Control.Exception (SomeException, try)
+import Data.Aeson
+  ( FromJSON,
+    parseJSON,
+    withObject,
+    (.:),
+  )
+import qualified Data.ByteString.Lazy as BL
+import Data.Functor ((<&>))
+import Data.List (intercalate)
+import Data.Maybe (fromJust, maybeToList)
+import Network.HTTP.Conduit (simpleHttp)
+import qualified Network.HTTP.Simple as HTTP
+import Path.IO (copyDirRecur, removeDirRecur)
+import StrongPath (Abs, Dir, File, Path', (</>))
+import qualified StrongPath as SP
+import StrongPath.Path (toPathAbsDir)
+import System.Directory (createDirectoryIfMissing)
+import Wasp.Cli.FileSystem (getUserCacheDir)
+
+type GithubRepo = (GithubRepoOwner, GithubRepoName)
+
+type GithubRepoOwner = String
+
+type GithubRepoName = String
+
+fetchFolderFromGithubRepoToDisk ::
+  forall d.
+  GithubRepo ->
+  String ->
+  Path' Abs (Dir d) ->
+  IO (Either String ())
+fetchFolderFromGithubRepoToDisk (githubOrg, repoName) pathToFolderInRepo destinationOnDisk = do
+  try
+    (withTempDir ".wasp-templates-temp" downloadFolderToDisk)
+    <&> either showException Right
+  where
+    downloadFolderToDisk :: Path' Abs (Dir d) -> IO ()
+    downloadFolderToDisk tempDir = do
+      downloadFile githubArchiveUrl tempDir archiveFileName >>= unpackArchive tempDir
+      copyRepoFolderToDestination tempDir pathToFolderInRepo destinationOnDisk
+      where
+        githubArchiveUrl = intercalate "/" ["https://github.com", githubOrg, repoName, "archive", archiveFileName]
+        archiveFileName = repoReference ++ ".tar.gz"
+
+    -- Which point in repo history to download (a branch or commit hash)
+    repoReference :: String
+    repoReference = "main"
+
+    downloadFile :: String -> Path' Abs (Dir d) -> String -> IO (Path' Abs (File f))
+    downloadFile downloadUrl destinationDir fileName = do
+      let destinationPath = destinationDir </> fromJust (SP.parseRelFile fileName)
+      simpleHttp downloadUrl >>= BL.writeFile (SP.fromAbsFile destinationPath)
+      return destinationPath
```

Comment:
It feels here like this function could have easily received the destinationPath already assembled, instead of assembling it on its own and then returning it. Seems like a responsibility that it got accidentally by code evolution, not because it should have it really. Yeah, I would assemble it outside and let it receive one argument less, also wouldn't return anything.